src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php

```diff
@@ -184,7 +184,10 @@ protected function isAllowedAttribute($classOrObject, string $attribute, ?string
                 self::$isWritableCache[$class.$attribute] = true;
             } else {
                 self::$isWritableCache[$class.$attribute] = $this->propertyInfoExtractor->isWritable($class, $attribute)
-                    || (($writeInfo = $this->writeInfoExtractor->getWriteInfo($class, $attribute)) && PropertyWriteInfo::TYPE_NONE !== $writeInfo->getType());
+                || (
+                    ($writeInfo = $this->writeInfoExtractor->getWriteInfo($class, $attribute)) && PropertyWriteInfo::TYPE_NONE !== $writeInfo->getType()
+                    && (PropertyWriteInfo::TYPE_PROPERTY !== $writeInfo->getType() || PropertyWriteInfo::VISIBILITY_PUBLIC === $writeInfo->getVisibility())
+                );
```

Comment:
Wouldn't this make more sense? The visibility check should have already been done on L186, isn't it?

```suggestion
                || (($writeInfo = $this->writeInfoExtractor->getWriteInfo($class, $attribute)) && !\in_array($writeInfo->getType(), [PropertyWriteInfo::TYPE_NONE, PropertyWriteInfo::TYPE_PROPERTY], true));
```

Reply:
I agree with you, this is much better!